### DISCUSSION TITLE
Dynamic display on buttons

### DISCUSSION THREAD

**Discussion by wpietri** [streamdeck-linux-gui/streamdeck-linux-gui#5] [Ideas / Feature Requests]
As requested, I'm moving this over [from the previous repo](https://github.com/timothycrosley/streamdeck-ui/issues/257).

I love this project! I've only had my Stream Deck for a couple of days, but it's already making my life better by letting me do common things instantly. The current release is great and already lets me do a lot of what I want. But it would be even better if there were a way to do buttons that change. Three examples:

1.    I have the Elgato Key Light, which I use during video calls. Right now I have two buttons, on and off. But what I'd rather have is one button that shows the current state of the light and pressing it does different things. E.g., if it's off, pressing it turns it on. If it's on and I hold it down, the brightness cycles.
 2.   I have a security camera that can detect motion. I already have a button that opens up the security cam interface. But I'd love if it I could have a motion alert change the image on the button. E.g., doing a short animation and then ending up with a red notification marker on top of the icon.
 3.   I'd love to be able to use the buttons for small graphs that update every 0.25 seconds or so. E.g. system load, website traffic levels. With then the button opening the relevant monitoring interface.

I could think of a few ways to do this. Maybe the simplest is a Python plugin system such that I subclass an object that receives keyup/keydown events and that can call a method to update the image. Or I could also imagine the Stream Deck UI app launching processes that talk a simple protocol over a socket.

**Comment by Danie10** [comment]
This may have to be spilt on a per-functionality basis? But the possibilities are endless with timers, time, and maybe also returning something to indicate whether something is turned on or off.

**Comment by Danie10** [comment]
@wpietri Wondering if this request should not rather have the subject "Dynamic Display on Buttons" so it better reflects what the outcome is, as I think many will support it?

**Reply by wpietri** [reply]
Done!

**Comment by sftim** [comment]
> a simple protocol over a socket

I'd like the integration to work via DBus.

**Example 1:** “on air” indicator
- pressing the button does nothing
- some other software uses DBus to update the state of a Stream Deck button
  (for example, the bottom left corner is always reserved as an “On Air” light)

**Example 2:** a button that controls the room lighting
- pressing the button commands the light to toggle (or turn on, or whatever); the Stream Deck is configured to send a DBus signal when that button is pressed
- some other software uses DBus to update the state of one or more Stream Deck buttons
  (for example, when the light is on, the “toggle”, ”brightness down” and ”brightness up” buttons all change appearance)

**Example 3:** a button set for video scenes
- pressing a transition button (“cut”, “dissolve”) runs a command to do a transition between the active source and the target
- pressing a source button (column 1 or column 5) runs a command to select a source
- some other software highlights the selected sources in each of the two columns
  when a transition happens, the target source becomes the active source (and is marked as such)
  when a transition happens, the active source becomes the target source (and is marked as such)

I am not sure whether the changes should be pushed via DBus or whether streamdeck-linux-gui should be able to subscribe to a value. Ideally, we'd support both.

For a mute button, the tool could subscribe to existing muted / unmuted notifications via DBus.

**Comment by coolapso** [comment]
Believe this feature will depend on this one: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/26

They can also be done simultaneously, but the way I see it https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/26 is the base / core of this feature.

**Reply by wpietri** [reply]
I could imagine it working either way.

Some of our examples above don't need state kept in the streamdeck-linux-gui code. E.g., for the lighting, the true state is stored in the lamp. For the on-air indicator, the state is in the casting software. For the graph displays, the state is also purely external.

I also think the state can be relatively rich. The on-air light is just a boolean, with just two fixed icons. My lighting is going to be on/off plus brightness and color temperature, so a boolean and two floats. The graph displays could be quite complicated. So if you're imagine offering transient/shadow state as a way of seeing whether we need to update the button, it should be pretty flexible.

So for my needs, I'd be happy with a pretty raw API that doesn't have any state storage at all. If we're going with the Python plugin approach, then I'd be happy with an instantiated object that was purely event driven: on_start, key_down, key_up, on_tick, on_end. Permanent state would live in the world; transient state would live in the object's instance variables. On events, new bitmaps would get sent off to the buttons.

I could imagine that on top of that others might want richer APIs, where transient state is managed for them. But that could be left out of the core app and instead be separate plugin code or even just examples for plugin developers to start from.

**Comment by bioxz** [comment]
I'm not sure if this fits better in this discussion or in #26, but as I had the same requirement I used bash scripts and streamdeckc as a workaround. 

First, I've added a script to my users autostart which reads the state of what I want to have shown on the buttons and changes the icon accordingly:

```
SPEAKER_MUTE=`pactl get-sink-mute @DEFAULT_SINK@ | awk '{print $2}'`

if [ "$SPEAKER_MUTE" = "yes" ]; then
	streamdeckc -a SET_ICON -b 1 --icon=/usr/share/icons/breeze-dark/status/24/audio-off.svg
else
	streamdeckc -a SET_ICON -b 1 --icon=/usr/share/icons/breeze-dark/status/24/audio-on.svg
fi
```
Next, the script was extended by a toggle functionality in front of the icon change which will a) be called from streamdeck_ui and b) can be called from outside triggers:

```
if [ "$1" = "toggle_speaker_mute" ]; then
	pactl set-sink-mute @DEFAULT_SINK@ toggle
fi
```

This works quite well for me, currently with two disadvantages: streamdeckc does not allow to set the icon background, which is only a minor problem and can most likely easily be fixed, and running the script actually takes some time and resources for each call of streamdeckc. Having such a feature integrated would obviously help a lot, as we can have access to the internal states and do not have to set an unchanged icon. Additionally you have to find ways to call the script of the state gets updated from somewhere else or use a timer, which is a bit problematic due to the load caused.

This can solve point 2 from the original post, partly point 1 (keeping pressed is not possible). Point 3 would work, but is resource intensive and 0.25 second might not work reliably.

The full script looks like this and is called during login (after steamdeck_ui) without parameters to set my audio icons to the current state. Then the buttons call the script with "<script> toggle_speaker_mute" and "<script> toggle_mic_mute".

```
#!/bin/sh

if [ "$1" = "toggle_speaker_mute" ]; then
	pactl set-sink-mute @DEFAULT_SINK@ toggle
fi

if [ "$1" = "toggle_mic_mute" ]; then
	pactl set-source-mute @DEFAULT_SOURCE@ toggle
fi


SPEAKER_MUTE=`pactl get-sink-mute @DEFAULT_SINK@ | awk '{print $2}'`

if [ "$SPEAKER_MUTE" = "yes" ]; then
	streamdeckc -a SET_ICON -b 1 --icon=/usr/share/icons/breeze-dark/status/24/audio-off.svg
else
	streamdeckc -a SET_ICON -b 1 --icon=/usr/share/icons/breeze-dark/status/24/audio-on.svg
fi

MIC_MUTE=`pactl get-source-mute @DEFAULT_SOURCE@ | awk '{print $2}'`

if [ "$MIC_MUTE" = "yes" ]; then
	streamdeckc -a SET_ICON -b 2 --icon=/usr/share/icons/breeze-dark/status/24/mic-off.svg
else
	streamdeckc -a SET_ICON -b 2 --icon=/usr/share/icons/breeze-dark/status/24/mic-on.svg
fi
```
